crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs

```diff
@@ -318,6 +319,57 @@ impl TEdgeTomlVersion {
     }
 }
 
+pub trait AppendRemoveItem {
+    type Item;
+
+    fn append(current_value: &mut Option<Self::Item>, new_value: Self::Item);
+
+    fn remove(current_value: &mut Option<Self::Item>, remove_value: Self::Item);
+}
+
+impl<T: Any + std::cmp::PartialEq> AppendRemoveItem for T {
+    type Item = T;
+
+    fn append(current_value: &mut Option<Self::Item>, new_value: Self::Item) {
+        if let Some(current_value) = current_value {
+            let ty_curr = current_value as &mut dyn Any;
+            let ty_new = &new_value as &dyn Any;
+
+            if let (Some(current), Some(new)) = (
+                ty_curr.downcast_mut::<TemplatesSet>(),
+                ty_new.downcast_ref::<TemplatesSet>(),
```

Comment:
I'm not convinced by using reflection here, partially because it means if we add more array-like configurations, we also have to remember this exists and cope appropriately. I'm also a bit wary of using reflection as it's based on the exact type used (so if the input was a `TemplatesSet` in the form of some dynamic trait object, we wouldn't be able to downcast to `TemplatesSet`).

There isn't any reason I can see why we need to have a generic implementation here, so this may have been a misunderstanding of my previous comments?

Reply:
The implementation cannot be specific to `c8y.smartrest.templates` and the `TemplateSet` type.

One should be able to use that for any setting that is a list, e.g. ` az.topics`

```
$ tedge config get az.topics                             
["te/+/+/+/+/m/+", "te/+/+/+/+/e/+", "te/+/+/+/+/a/+", "te/+/+/+/+/status/health"]

$ tedge config add az.topics azure

$ tedge config get az.topics           
["azure"]
```

I was expecting `"te/+/+/+/+/m/+", "te/+/+/+/+/e/+", "te/+/+/+/+/a/+", "te/+/+/+/+/status/health","azure"]` and not `["azure"]`

Reply:
> I'm not convinced by using reflection here, partially because it means if we add more array-like configurations, we also have to remember this exists and cope appropriately. I'm also a bit wary of using reflection as it's based on the exact type used (so if the input was a `TemplatesSet` in the form of some dynamic trait object, we wouldn't be able to downcast to `TemplatesSet`). 
> 
> There isn't any reason I can see why we need to have a generic implementation here, so this may have been a misunderstanding of my previous comments?

I thought it would be easier to maintain the generic implementation. In first implementation you had to fill macro with all the types (and then remember to append it if new type was added). But you are right, I haven't thought about the dynamic trait object. With additional array-like configurations you could modify current solution to check all array types with `is::<TemplateSet>()` in if..else blocks but I see it can get pretty clunky over time. I will reverse current solution to previous one with macro.

> The implementation cannot be specific to `c8y.smartrest.templates` and the `TemplateSet` type.
> 
> One should be able to use that for any setting that is a list, e.g. ` az.topics`

AFAIK, the `TemplateSet` is the only array type field in `tedge config` (All topic fields use `TemplateSet`). Apparently, there is an issue with default value that it is not present when I am checking current value of the fields, therefore instead of appending, it replace the value. I did not notice that with `c8y.smartrest.templates` because it is empty by default. I need to investigate how exactly default value is assigned.

Reply:
>> The implementation cannot be specific to `c8y.smartrest.templates` and the `TemplateSet` type.
>> 
>> One should be able to use that for any setting that is a list, e.g. ` az.topics`

> AFAIK, the `TemplateSet` is the only array type field in `tedge config` (All topic fields use `TemplateSet`). Apparently, there is an issue with default value that it is not present when I am checking current value of the fields, therefore instead of appending, it replace the value. I did not notice that with `c8y.smartrest.templates` because it is empty by default. I need to investigate how exactly default value is assigned.

You are correct. There is an issue *only* when the list has not been explicitly set by the user, but provided as a default. This is working fine once a first list has been set by the user.

```
$ sudo tedge config set az.topics 'foo,bar'
$ tedge config get az.topics
["foo", "bar"]
$ sudo tedge config add az.topics 'azure'  
$ tedge config get az.topics             
["foo", "bar", "azure"]
```

I have been also positively surprised to see that *append* is working too:

```
$ sudo tedge config add az.topics 'xyz,abc'
$ tedge config get az.topics               
["foo", "bar", "azure", "xyz", "abc"]
```